utility_analysis_new/combiners.py

```diff
@@ -0,0 +1,135 @@
+
+
+@dataclass
+class CountUtilityAnalysisMetrics:
+    """ Stores metrics for the count utility analysis.
+
+    Args:
+        count: the number of contributions per a partition.
+        per_partition_contribution_error: the amount of contribution error in a partition.
+        expected_cross_partition_error: the estimated amount of error across partitions.
+        var_cross_partition_error: the variance of the contribution bounding error.
+        std_cb: the standard deviation of the var_cross_partition_error.
+        std_noise: the noise standard deviation for DP count.
+        noise_kind: the type of noise used.
+    """
+    count: int
+    per_partition_contribution_error: int
+    expected_cross_partition_error: float
+    var_cross_partition_error: float
+    std_cb: float
+    std_noise: float
+    noise_kind: pipeline_dp.NoiseKind
+
+
+@dataclass
+class AccumulatorMetrics:
+    count: int
+    per_partition_contribution_error: int
+    expected_cross_partition_error: float
+    var_cross_partition_error: float
+
+
+class UtilityAnalysisCountCombiner(pipeline_dp.Combiner):
+    """ A combiner for utility analysis counts. """
+    AccumulatorType = AccumulatorMetrics
+
+    def __init__(self, params: pipeline_dp.combiners.CombinerParams):
+        self._params = params
+
+    def create_accumulator(self, data: Sized) -> AccumulatorType:
+        """ Create an accumulator for a user and a partition.
+
+        Args:
+            data: contains two columns; 1) a column with the total number of partitions a user contributed to
+            and 2) a column with a list of the user's contributions for a single partition.
+
+        Returns:
+            An accumulator with the count of contributions and the contribution error for this partition, and the
+            estimated error and variance across all of the partitions for this user.
+        """
+        if not data:
+            return AccumulatorMetrics(0, 0, 0, 0)
+        n_partitions, values = data
+        count = len(values)
+        max_per_partition = self._params.aggregate_params.max_contributions_per_partition
+        max_partitions = self._params.aggregate_params.max_partitions_contributed
+        per_partition_contribution_error = max(0, count - max_per_partition)
+        prob_keep_partition = min(1, max_partitions /
+                                  n_partitions) if n_partitions > 0 else 0
+        expected_cross_partition_error = per_partition_contribution_error * prob_keep_partition
+        var_cross_partition_error = per_partition_contribution_error**2 * prob_keep_partition * (
+            1 - prob_keep_partition)
+
+        return AccumulatorMetrics(count, per_partition_contribution_error,
+                                  expected_cross_partition_error,
+                                  var_cross_partition_error)
+
+    def merge_accumulators(self, acc1: AccumulatorType, acc2: AccumulatorType):
+        """ Merge two accumulators together additively.
```

Comment:
Nit: 1 line docstring is enough (i.e. no need Args: ...)

Reply:
Got it, removed to make it simpler. Thank you!